From 504fdff7fb2bbbdf717943928ddc69cfaa5cc9f2 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Fri, 17 Nov 2023 17:28:50 +0100
Subject: [PATCH 0089/1695] pta: stats: add commands to print clock and
 regulator trees

Add statistics PTA commands STATS_CMD_PRINT_DRIVERS_INFO to print
device drivers information on console. The implementation currently
allows to print the clock tree and the regulator tree to core console.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git cec3eac6326f03adc87331a163d9efdfcd675957 pta: stats: add commands to print clock and regulator trees]
Change-Id: I199ddf5cfdac79b3a226d9b805e0285f78f0ff9b
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
[etienne: cherry-picked from a5e75b7ea399137e60503bb6693e3cb03ef2b8eb]
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/pta/stats.c                | 26 ++++++++++++++++++++++++++
 lib/libutee/include/pta_stats.h | 10 ++++++++++
 2 files changed, 36 insertions(+)

diff --git a/core/pta/stats.c b/core/pta/stats.c
index 86dc3d193..22288ef16 100644
--- a/core/pta/stats.c
+++ b/core/pta/stats.c
@@ -3,6 +3,8 @@
  * Copyright (c) 2015, Linaro Limited
  */
 #include <compiler.h>
+#include <drivers/clk.h>
+#include <drivers/regulator.h>
 #include <kernel/pseudo_ta.h>
 #include <kernel/tee_time.h>
 #include <malloc.h>
@@ -173,6 +175,28 @@ static TEE_Result get_system_time(uint32_t type,
 	return TEE_SUCCESS;
 }
 
+static TEE_Result print_driver_info(uint32_t type, TEE_Param p[TEE_NUM_PARAMS])
+{
+	if (TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+			    TEE_PARAM_TYPE_NONE,
+			    TEE_PARAM_TYPE_NONE,
+			    TEE_PARAM_TYPE_NONE) != type)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	switch (p[0].value.a) {
+	case STATS_DRIVER_TYPE_CLOCK:
+		clk_print_tree();
+		break;
+	case STATS_DRIVER_TYPE_REGULATOR:
+		regulator_print_tree();
+		break;
+	default:
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
 /*
  * Trusted Application Entry Points
  */
@@ -192,6 +216,8 @@ static TEE_Result invoke_command(void *psess __unused,
 		return get_user_ta_stats(ptypes, params);
 	case STATS_CMD_GET_TIME:
 		return get_system_time(ptypes, params);
+	case STATS_CMD_PRINT_DRIVER_INFO:
+		return print_driver_info(ptypes, params);
 	default:
 		break;
 	}
diff --git a/lib/libutee/include/pta_stats.h b/lib/libutee/include/pta_stats.h
index 25ed58c69..f95d3acb8 100644
--- a/lib/libutee/include/pta_stats.h
+++ b/lib/libutee/include/pta_stats.h
@@ -81,4 +81,14 @@ struct pta_stats_ta {
  */
 #define STATS_CMD_GET_TIME		4
 
+/*
+ * STATS_CMD_PRINT_DRIVER_INFO - Print device drivers information to console
+ *
+ * [in]    value[0].a        Target driver, one of STATS_DRIVER_TYPE_*
+ */
+#define STATS_CMD_PRINT_DRIVER_INFO	5
+
+#define STATS_DRIVER_TYPE_CLOCK		0
+#define STATS_DRIVER_TYPE_REGULATOR	1
+
 #endif /*__PTA_STATS_H*/
-- 
2.47.2

